### DISCUSSION TITLE
Lathecode is great!

### DISCUSSION THREAD

**Discussion by Festivejelly** [kachurovskiy/nanoels#163] [General]
Just an appreciation post for lathecode. Its such a great tool makes more complex shapes a dream. Its very intuitive and is the icing on the cake for this controller.

https://youtu.be/PH6cTw_ml4M?si=eXcVM9Pg6dy2jgbW

**Comment by kachurovskiy** [comment]
Nicholas, glad it worked. Thanks for sharing!

**Reply by Festivejelly** [reply]
I managed to make a dozen parts in minutes it works super well.

All thats missing is boring operations, if I get a chance i'll see if I can implement it. But just as it is it saves me a ton of time. Generally I use drills and reamers anyway to get exact sizes but being able to do some boring using the lathe code would make that easier and cheaper as you wouldnt need a reamer for each size lol.

**Reply by kachurovskiy** [reply]
Yeah, inside operations aren't implemented yet - but for something as simple as boring it's possible to write a few lines of GCode by hand and send it using https://kachurovskiy.github.io/lathecode/ (modify GCode before send) or via https://kachurovskiy.github.io/nanoels/h4/sender.html

```
; GCode below is an untested example

G21 ; metric
G18 ; ZX plane
G90 ; absolute positioning
F200
X-3 ; absolute start X 
Z0 ; absolute start Z
G91 ; relative positioning

F10 ; slow cut speed
Z10 ; do first boring pass 10mm deep

F200 ; fast
X0.5 ; retract 0.5mm X
Z-10 ; retract Z

F10 ; slow
X-1 ; move in X for second boring pass
Z10 ; do the pass
```

**Reply by Festivejelly** [reply]
Ah yes good suggestion. With the sender I assume it sends it line by line, so I could tweak it manually. My use case for this is to make 100's of the same item so im willing to put in the effort for that. Ive got a set of quite small boring bars and grooving tools for internal operations.

Ive been enjoying going through the code and seeing how it all fits together.

Which made me think maybe I could have a crack at tool changing using the settings menu.

I generally use 3 tools:

T0 - Parting Tool
T1 - Boring Tool
T2 - Grooving Tool

T1 and T2 values should be relative to T0. Which means I only ever need to zero from T0. This assumes that the quick change tool post is repeatable, which for my purposes its accurate enough.

Since the settings menu is not currently used I might be able to put something in there. Like it could have something like "current tool: T0" then highlighting it with the arrow keys and pressing left or right will cycle the list of tools and apply its offset.

I think ill probably store the tools like this:

typedef struct
{
  const char *toolName;
  float *offsetZ;
  float *offsetX;
} Tool;

Tool tools[] = {
    {"T0", 0, 0},
    {"T1", 0, 2},
    {"T2", 3, 1}
    // Add more tools here
};

Other settings could easily be added in the same way too as the settings menu could be paginated and you could just scroll through it with the arrows.

I'll share a demo when its done. I guess once ive worked out then how to do it via the buttons I could add gcode support for it, and maybe have it go back to the start and pause while you change tools.
